opt: fix exponential blowup of semi-join ordering

DrewKimball opened the pull request:
Previously, a join tree with multiple `semi-join`s could cause an
exponential blowup during optimization due to interaction between
`CommuteSemiJoin` and `ReorderJoins`. Consider the following query:
```
SELECT * FROM xy
WHERE EXISTS (SELECT 1 FROM ab WHERE b = y)
AND EXISTS (SELECT 1 FROM ab WHERE b = y)
AND EXISTS (SELECT 1 FROM ab WHERE b = y);
```
Each `semi-join` in this join tree will be reordered, producing multiple
alternate join trees each with alternate `semi-joins`. At each step, the
new `semi-join`s are added to the group of the original. Since `ReorderJoins`
only matches the first expression of a group, it will not rematch on the
new `semi-joins`.

This is where `CommuteSemiJoin` becomes relevant. `CommuteSemiJoin` matches
every `semi-join` in the memo, and produces a new join group. `ReorderJoins`
then matches this new join group, and reorders it. In a tree with multiple
`semi-join`s, the inputs of such a group may include still more `semi-join`s,
each of which will itself engage in this cycle. So, the result is an
exponential blowup of memo expressions that may occur when a join tree has
multiple `semi-join`s.

This patch modifies `JoinOrderBuilder` to keep track of expressions which were
constructed as a result of reordering. The `alreadyOrdered` map is used to
prevent `ReorderJoins` from matching on these expressions.

Note that exploration rules which replace a join with a new join group (rather
than adding to the existing one) must use the `PreventReordering` method as
`CommuteSemiJoin` now does. This will prevent exponential blowups due to
interaction between the rule and `ReorderJoins`.

Release note: None

Fixes #59076

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59128)
<!-- Reviewable:end -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[pkg/sql/opt/memo/interner.go, line 509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MROLPUuETso5Gl6kGxB:-MROLPUvENS0w-ipK6Wr:b-5uetab) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ca2778a059e358216dc56faa7ef06fc91d75e/pkg/sql/opt/memo/interner.go#L509)):*
> ```Go
> }
> 
> func (h *hasher) HashJoinWasReordered(val JoinWasReordered) {
> ```

I'm not sure I like this solution, because it sets a new precedent. Before, we could always rely on every field being interned; now we may be surprised when some fields are and others are not.

Instead, I think we should consider removing this flag altogether from the Join privates and using a `wasReordered` map in the `JoinOrderBuilder` instead to track joins that have already been processed.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball)

___
*[pkg/sql/opt/memo/interner.go, line 509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MROLPUuETso5Gl6kGxB:-MROMDKR5syqDzVFTLLI:b40ek5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ca2778a059e358216dc56faa7ef06fc91d75e/pkg/sql/opt/memo/interner.go#L509)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'm not sure I like this solution, because it sets a new precedent. Before, we could always rely on every field being interned; now we may be surprised when some fields are and others are not.

Instead, I think we should consider removing this flag altogether from the Join privates and using a `wasReordered` map in the `JoinOrderBuilder` instead to track joins that have already been processed.
</blockquote></details>

Problem is, that would invalidate my fix, since the inner join resulting from `CommuteSemiJoin` wouldn't be in the `wasReordered` map, and we'd be back to square one. I guess maybe `CommuteSemiJoin` could add to the map if it itself is part of it?

Another option would be to intern the `WasReordered` field after all. It would lead to some duplicate effort, but it wouldn't be _too_ bad.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
Maybe instead of storing pointers to joins, we store pointers to join privates. That might do the trick.

DrewKimball commented:
Ah, never mind. Joins don't store `JoinPrivate` pointers.

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[pkg/sql/opt/memo/interner.go, line 509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MROLPUuETso5Gl6kGxB:-MRQCjU-DhVmwpvXgdU7:b-hs9fzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ca2778a059e358216dc56faa7ef06fc91d75e/pkg/sql/opt/memo/interner.go#L509)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Problem is, that would invalidate my fix, since the inner join resulting from `CommuteSemiJoin` wouldn't be in the `wasReordered` map, and we'd be back to square one. I guess maybe `CommuteSemiJoin` could add to the map if it itself is part of it?

Another option would be to intern the `WasReordered` field after all. It would lead to some duplicate effort, but it wouldn't be _too_ bad.
</blockquote></details>

I don't think this PR sets the `WasReordered` bit in the `CommuteSemiJoin` operator, does it? Don't both solutions have the same issue?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball)

___
*[pkg/sql/opt/memo/interner.go, line 509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MROLPUuETso5Gl6kGxB:-MRQPuxoD00LIvNr1I9e:b-kk7tsh) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ca2778a059e358216dc56faa7ef06fc91d75e/pkg/sql/opt/memo/interner.go#L509)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I don't think this PR sets the `WasReordered` bit in the `CommuteSemiJoin` operator, does it? Don't both solutions have the same issue?
</blockquote></details>

It sets the `WasReordered` bit for any joins that are the result of a reordering, and `CommuteSemiJoin` reuses the matched `semi-join`'s join private.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[pkg/sql/opt/memo/interner.go, line 509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MROLPUuETso5Gl6kGxB:-MRQdbLG9I-jHN-cZWIj:b-sriqgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ca2778a059e358216dc56faa7ef06fc91d75e/pkg/sql/opt/memo/interner.go#L509)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It sets the `WasReordered` bit for any joins that are the result of a reordering, and `CommuteSemiJoin` reuses the matched `semi-join`'s join private.
</blockquote></details>

Ah I see. I think it'd be fine to create a custom function to create a join that should not be reordered. It would call into the join order builder to add itself to the `wasReordered` map


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball)

___
*[pkg/sql/opt/memo/interner.go, line 509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MROLPUuETso5Gl6kGxB:-MRQezm07FkLOKtM_28K:bbh87w1) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ca2778a059e358216dc56faa7ef06fc91d75e/pkg/sql/opt/memo/interner.go#L509)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Ah I see. I think it'd be fine to create a custom function to create a join that should not be reordered. It would call into the join order builder to add itself to the `wasReordered` map
</blockquote></details>

Alright, that sounds good to me.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball)

___
*[pkg/sql/opt/memo/interner.go, line 509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MROLPUuETso5Gl6kGxB:-MRT4u4vAZXyf8QvsjlK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/989ca2778a059e358216dc56faa7ef06fc91d75e/pkg/sql/opt/memo/interner.go#L509)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Alright, that sounds good to me.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice clean fix!

> Reviewed 8 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-:-MRVlWVB856a2iUc8eQg:b-vfulrd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I wanted to suggest adding a regression test (with the query in the issue) in `xform` where we just verify that the number of rules applied is not ridiculous, but we can't do that easily now; filed https://github.com/cockroachdb/cockroach/issues/59192.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-:-MRVr3Tz3zPyxkgyBmiM:b-7mn074)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @DrewKimball)

___
*[pkg/sql/opt/xform/join_order_builder.go, line 789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MRWlMTz0S24LRDzG_cu:-MRWlMTz0S24LRDzG_cv:bmkor1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fec34ede4e34b422b788ccce3c83c0db991a84c/pkg/sql/opt/xform/join_order_builder.go#L789)):*
> ```Go
> 	// re-matching on it. Don't add to the map if this is the originally matched
> 	// root of the join tree, because we want to match on any new join groups
> 	// which are generated from it.
> ```

I don't understand this, can you give example where we don't want to add?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @rytaft)

___
*[pkg/sql/opt/xform/join_order_builder.go, line 789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MRWlMTz0S24LRDzG_cu:-MRXB3mr0bE1YCvBzQtE:b-c45me7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fec34ede4e34b422b788ccce3c83c0db991a84c/pkg/sql/opt/xform/join_order_builder.go#L789)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I don't understand this, can you give example where we don't want to add?
</blockquote></details>

I've added an explanation here - let me know if it's sufficient.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball, @DrewKimball, and @rytaft)

___
*[pkg/sql/opt/xform/join_order_builder.go, line 789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MRWlMTz0S24LRDzG_cu:-MRXEBQ667J1Wcx6ru9Y:b7k0wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fec34ede4e34b422b788ccce3c83c0db991a84c/pkg/sql/opt/xform/join_order_builder.go#L789)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I've added an explanation here - let me know if it's sufficient.
</blockquote></details>

It seems simpler to:
1. Add joins that create a new group to `alreadyOrderedMap`.
2. Don't add joins that are added to an existing group to `alreadyOrderedMap`. These aren't matched anyway.
3. Make the `PreventReordering` function always add its single input (only need 1 input) to `alreadyOrderedMap`.

Is there some reason that wouldn't work? It'd collapse this new code down to something simpler.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @rytaft)

___
*[pkg/sql/opt/xform/join_order_builder.go, line 789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MRWlMTz0S24LRDzG_cu:-MRXOblr01-qxTAus7c4:b-7w7bm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fec34ede4e34b422b788ccce3c83c0db991a84c/pkg/sql/opt/xform/join_order_builder.go#L789)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

It seems simpler to:
1. Add joins that create a new group to `alreadyOrderedMap`.
2. Don't add joins that are added to an existing group to `alreadyOrderedMap`. These aren't matched anyway.
3. Make the `PreventReordering` function always add its single input (only need 1 input) to `alreadyOrderedMap`.

Is there some reason that wouldn't work? It'd collapse this new code down to something simpler.
</blockquote></details>

Well, it would prevent reordering on the `inner-join`s created from `semi-join`s. If that's ok, I'm happy to make the change, since it is more simple. 

Note that doing this would change the plan for TPCH Q20 (though I think the plan might actually be faster), as well as a number of other tests.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball, @DrewKimball, and @rytaft)

___
*[pkg/sql/opt/xform/join_order_builder.go, line 789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MRWlMTz0S24LRDzG_cu:-MRXbXJSAE_A1celLac8:budtzdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fec34ede4e34b422b788ccce3c83c0db991a84c/pkg/sql/opt/xform/join_order_builder.go#L789)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Well, it would prevent reordering on the `inner-join`s created from `semi-join`s. If that's ok, I'm happy to make the change, since it is more simple. 

Note that doing this would change the plan for TPCH Q20 (though I think the plan might actually be faster), as well as a number of other tests.
</blockquote></details>

But I thought your change would also prevent reordering on those inner joins. `CommuteSemiJoin` should only match on a join that's already been ordered, which means the `InnerJoin` it creates will also be marked as "already ordered". Am I misunderstanding something?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @rytaft)

___
*[pkg/sql/opt/xform/join_order_builder.go, line 789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MRWlMTz0S24LRDzG_cu:-MRXcgdv6kXnAdmDLuX3:b-emlyxw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fec34ede4e34b422b788ccce3c83c0db991a84c/pkg/sql/opt/xform/join_order_builder.go#L789)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

But I thought your change would also prevent reordering on those inner joins. `CommuteSemiJoin` should only match on a join that's already been ordered, which means the `InnerJoin` it creates will also be marked as "already ordered". Am I misunderstanding something?
</blockquote></details>

The exception is for the original join tree. So, an `inner join` created from a `semi join` that was in the original normalized tree will be reordered, but an `inner join` from the reordered version of that `semi join` would not be reordered.

It's a way to try and generate many of the same plans as before without exposing ourselves to the exponential blowup problem. 


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball, @DrewKimball, and @rytaft)

___
*[pkg/sql/opt/xform/join_order_builder.go, line 789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MRWlMTz0S24LRDzG_cu:-MRXkrm42_nAe4qH7STX:b-yf5y4z) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fec34ede4e34b422b788ccce3c83c0db991a84c/pkg/sql/opt/xform/join_order_builder.go#L789)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

The exception is for the original join tree. So, an `inner join` created from a `semi join` that was in the original normalized tree will be reordered, but an `inner join` from the reordered version of that `semi join` would not be reordered.

It's a way to try and generate many of the same plans as before without exposing ourselves to the exponential blowup problem. 
</blockquote></details>

Hm, I think I need to actually step through this, because I don't see why that || expression would accomplish that. `ordered` is a new join being added to an existing group. Therefore, it should never be the first expression in the group, and so it'd always be marked as "already ordered".


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @rytaft)

___
*[pkg/sql/opt/xform/join_order_builder.go, line 789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MRWlMTz0S24LRDzG_cu:-MRXm0Q67SiJ5MCk5y_L:b22sz7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fec34ede4e34b422b788ccce3c83c0db991a84c/pkg/sql/opt/xform/join_order_builder.go#L789)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Hm, I think I need to actually step through this, because I don't see why that || expression would accomplish that. `ordered` is a new join being added to an existing group. Therefore, it should never be the first expression in the group, and so it'd always be marked as "already ordered".
</blockquote></details>

Ah, I see the problem here. `JoinOrderBuilder` will reproduce the originally matched join, and relies on the interner to prevent the duplicate from actually being added to the memo. Since the original joins are all added into the `plans` map, the duplicate is added to the same group as the original.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @rytaft)

___
*[pkg/sql/opt/xform/join_order_builder.go, line 789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MRWlMTz0S24LRDzG_cu:-MRXn1rQC2YvfEuZ9LcI:bwderkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fec34ede4e34b422b788ccce3c83c0db991a84c/pkg/sql/opt/xform/join_order_builder.go#L789)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Ah, I see the problem here. `JoinOrderBuilder` will reproduce the originally matched join, and relies on the interner to prevent the duplicate from actually being added to the memo. Since the original joins are all added into the `plans` map, the duplicate is added to the same group as the original.
</blockquote></details>

Maybe I should clarify this in the comment?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @rytaft)

___
*[pkg/sql/opt/xform/join_order_builder.go, line 789 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-MRWlMTz0S24LRDzG_cu:-MRXrAxtBlyXIHjYwABQ:b-jkw25r) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fec34ede4e34b422b788ccce3c83c0db991a84c/pkg/sql/opt/xform/join_order_builder.go#L789)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Maybe I should clarify this in the comment?
</blockquote></details>

Actually, now that I think about it more, the root check isn't necessary; if we're in `addJoin` and `ordered` is the first expression of the group, this must be one of the originally matched joins.


<!-- Sent from Reviewable.io -->

DrewKimball committed:
opt: fix exponential blowup of semijoin ordering

Previously, a join tree with multiple `semi-join`s could cause an
exponential blowup during optimization due to interaction between
`CommuteSemiJoin` and `ReorderJoins`. Consider the following query:
```
SELECT * FROM xy
WHERE EXISTS (SELECT 1 FROM ab WHERE b = y)
AND EXISTS (SELECT 1 FROM ab WHERE b = y)
AND EXISTS (SELECT 1 FROM ab WHERE b = y);
```
Each `semi-join` in this join tree will be reordered, producing multiple
alternate join trees each with alternate `semi-joins`. At each step, the
new `semi-join`s are added to the group of the original. Since `ReorderJoins`
only matches the first expression of a group, it will not rematch on the
new `semi-joins`.

This is where `CommuteSemiJoin` becomes relevant. `CommuteSemiJoin` matches
every `semi-join` in the memo, and produces a new join group. `ReorderJoins`
then matches this new join group, and reorders it. In a tree with multiple
`semi-join`s, the inputs of such a group may include still more `semi-join`s,
each of which will itself engage in this cycle. So, the result is an
exponential blowup of memo expressions that may occur when a join tree has
multiple `semi-join`s.

This patch modifies `JoinOrderBuilder` to avoid matching on any joins that
are the result of reordering, whether or not the join is part of a new memo
group. This will prevent exponential blowups due to interaction between the
rule and `ReorderJoins`.

Release note: None

Fixes #59076

DrewKimball commented:
After discussion with @andy-kimball I've simplified the change to be a bit less fragile / difficult to understand.

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59128#-:-MRYuPM097cCATnb8JYK:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @andy-kimball and @rytaft)



<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+

andy-kimball commented:
bors r+

RaduBerinde commented:
We should backport to 20.2, I can do that unless you were planning to.

DrewKimball commented:
I should be able to do it later today.
